File: services/python/starlette/src/routes/handlers/delete.py
import structlog
from uuid import UUID
from starlette.requests import Request
from starlette.responses import Response

from app.state import AppState
from app.task import Task
from routes.errors import send_error, AppErrors


async def delete_handler(request: Request) -> Response:
    logger = structlog.get_logger()

    id: UUID = request.path_params["id"]
    state: AppState = request.app.state.app_state

    async with request.app.state.tasks_lock:
        task: Task | None = state.tasks.pop(id, None)

        if task is None:
            logger.warn(
                "Task not found",
                id=str(id),
                method=request.method,
                path=request.url.path,
            )
            return send_error(AppErrors.TaskNotFound)

    logger.info(
        "Removed task",
        id=str(id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method=request.method,
    )

    return Response(status_code=204)

File: services/go/chi/routes/handlers/delete.go
package handlers

import (
	"app"
	"chi/routes"
	"log/slog"
	"net/http"
	"strings"
)

func DeleteHandler(state *app.AppState) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := ParseID(w, r)
		if err != nil {
			return
		}

		state.Mu.Lock()
		task, ok := state.Tasks[id]

		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", r.Method, "path", r.URL.Path, "id", id.String())
			routes.AppErrors.TaskNotFound.Error(w)
			return
		}

		delete(state.Tasks, id)

		state.Mu.Unlock()

		slog.Info(
			"Removed task",
			"id", id.String(),
			"secret", len(task.Secret),
			"operation", strings.ToLower(string(task.Operation)),
			"method", r.Method,
		)

		w.WriteHeader(http.StatusNoContent)
	}
}
